consensus/blocknode.go

```diff
@@ -72,11 +77,35 @@ func (s *State) addBlockToTree(b Block) (newNode *BlockNode) {
 	}
 	newNode.setTarget()
 
-	// Add the node to the block map and the list of its parents children.
+	// Add the node to the block map and update the list of its parents
+	// children.
 	s.blockMap[b.ID()] = newNode
 	parentNode.Children = append(parentNode.Children, newNode)
 
-	// TODO TODO TODO: reconnect any orphans that were liberated by this block.
+	if s.heavierFork(newNode) {
+		err = s.forkBlockchain(newNode)
+		if err != nil {
+			return
+		}
+	}
+
+	// Reconnect all orphans that now have a parent.
+	childMap, exists := s.missingParents[b.ID()]
+	if exists {
+		for _, child := range childMap {
+			// This is a recursive call, which means someone could make it take
+			// a long time by giving us a bunch of false orphan blocks with the
+			// same parent. We have to check them all anyway though, it just
+			// allows them to cluster the processing. Utimately I think
+			// bandwidth will be the bigger issue, we'll reject bad orphans
+			// before we verify signatures.
+			_ = s.addBlockToTree(child)
+			// There's nothing we can really do about this error, because it
+			// doesn't reflect a problem with the current block. It means that
+			// someone managed to give us orphans with errors.
+		}
+		delete(s.missingParents, b.ID())
+	}
```

Comment:
`exists` is technically unnecessary here, because a `range` over a nil map is a no-op, and a `delete` on a nil map is too.
```go
for _, child := range s.missingParents[b.ID()] {
    s.addBlockToTree(child)
}
delete(s.missingParents, b.ID())
```